src/pql/query_sql.py

```diff
@@ -0,0 +1,72 @@
+import typing
+
+import pandas as pd
+from pandasql import sqldf
+
+from src.pql.exceptions import UserQueryError, ParserNotFound, ParseDataError
+
+
+def to_df(data: typing.Any, parser: typing.Optional[str]) -> pd.DataFrame:
+    """Convert data to pd.DataFrame.
+
+    Args:
+        data (typing.Any): The data to convert to pd.DataFrame.
+        parser (str): The parser required to convert data to pd.DataFrame.
+
+    Returns:
+        pd.DataFrame: the data in pd.DataFrame format
+    """
+    try:
+        if parser == "json":
+            return pd.json_normalize(data)
+        if parser == "list":
+            return pd.DataFrame([data])
+        if parser is None:
+            return data
+        else:
+            raise ParserNotFound(f"parser not found - {parser}")
+    except ParserNotFound:
+        raise
+    except Exception:
+        raise ParseDataError(f"failed to parse data {data} using parser {parser}")
+
+
+async def execute_sql_query(
+    request_data: dict, query: str, result: typing.Optional[bool]
+) -> typing.Any:
+    """Execute user defined query against request data.
+
+    Args:
+        request_data: dictionary containing request data formatted as pd.DataFrame.
+        query: user defined query to run against request data.
+        result: index of the result.
+
+    Returns:
+        typing.Any: sql query result
+    """
+    try:
+        if result:
+            return sqldf(query, request_data).iloc[0, 0]
```

Comment:
This is some dark magic. :tophat: Please explain what `result` field in JSON does. The docstring says: `index of the result`.

Reply:
Dark magic indeed 🎩 

This logic assumes that the desired result of the query response is located at column 0 and row 0 which is equivalent to iloc[0, 0] and therefore extracts the value at this location.  The reason we do not extract the result by default is due to the following.

The sql query logic can be executed from two points in the code:
1. As a pipeline step
2. As an aggregation step

If the sql query is executed as a pipeline step then the query response may not be the final result.  It could instead be an intermediary representation of the source data that needs to be further processed via PQL aggregation.  Having the result switch allows the user to define whether the query response is the final result or if it needs to be further processed downstream by the aggregation method.  

The reason the docstring reads as you mention is because previously I allowed the user to specify the indexes (column + row) of the desired result instead of assuming the result was located at [0,0].  I'm uncertain if the index specification or [0,0] assumption implementation is better, any thoughts?  I'll update the docstring  to be more reflective of it's currently utility.

Reply:
Thank you for your explanation. I would just leave it as it right now, we will see how it will be used and we can update it later.